src/DataProtection/DataProtection/src/KeyManagement/KeyRingProvider.cs

```diff
@@ -292,6 +311,146 @@ internal IKeyRing GetCurrentKeyRingCore(DateTime utcNow, bool forceRefresh = fal
         }
     }
 
+    private IKeyRing GetCurrentKeyRingCoreNew(DateTime utcNow, bool forceRefresh)
+    {
+        Debug.Assert(utcNow.Kind == DateTimeKind.Utc);
+
+        // The 99% and perf-critical case is that there is no task in-flight and the cached
+        // key ring is valid.  We do what we can to avoid unnecessary overhead (locking,
+        // context switching, etc) on this path.
+
+        CacheableKeyRing? existingCacheableKeyRing = null;
+
+        // Can we return the cached keyring to the caller?
+        if (!forceRefresh)
+        {
+            existingCacheableKeyRing = Volatile.Read(ref _cacheableKeyRing);
+            if (CacheableKeyRing.IsValid(existingCacheableKeyRing, utcNow))
+            {
+                return existingCacheableKeyRing.KeyRing;
+            }
+        }
+
+        // If work kicked off by a previous caller has completed, we should use those results
+        // We check this outside the lock to reduce contention in the common case (no task).
+        // Logically, it would probably make more sense to check this before checking whether
+        // the cache is valid - there could be a newer value available - but keeping that path
+        // fast is more important.  The next forced refresh or cache expiration will cause the
+        // new value to be picked up.
+        var existingTask = Volatile.Read(ref _cacheableKeyRingTask);
+        if (existingTask is not null && existingTask.IsCompleted)
+        {
+            var taskKeyRing = GetKeyRingFromCompletedTask(existingTask, utcNow); // Throws if the task failed
+            if (taskKeyRing is not null)
+            {
+                return taskKeyRing;
+            }
+        }
+
+        // The cached keyring hasn't been created or must be refreshed. We'll allow one thread to
+        // create a task to update the keyring, and all threads will continue to use the existing cached
+        // keyring while the first thread performs the update. There is an exception: if there
+        // is no usable existing cached keyring, all callers must block until the keyring exists.
+        lock (_cacheableKeyRingLockObj)
+        {
+            // Update existingTask, in case we're not the first to acquire the lock
+            existingTask = Volatile.Read(ref _cacheableKeyRingTask);
+            if (existingTask is null)
+            {
+                // If there's no existing task, make one now
+                // PERF: Closing over utcNow substantially slows down the fast case (valid cache) in micro-benchmarks
+                // (closing over `this` for CacheableKeyRingProvider doesn't seem impactful)
+                existingTask = Task.Factory.StartNew(
+                    utcNow => CacheableKeyRingProvider.GetCacheableKeyRing((DateTime)utcNow!),
```

Comment:
I don't think the null suppressor is needed here?

Reply:
I think it is because that's referring to the parameter and not the local we know is non-null.  Usually, I rename such parameters for clarity, but I couldn't think of an alternative name.

Reply:
Confirmed that the type of `utcNow` is `object?`, necessitating the case and the null suppressor.